File: dashpilot-android/app/src/main/java/com/softwiredtech/dashpilot/viewmodel/ConnectionViewModel.kt
package com.softwiredtech.dashpilot.viewmodel

import android.Manifest
import android.annotation.SuppressLint
import android.bluetooth.BluetoothDevice
import android.bluetooth.BluetoothManager
import android.bluetooth.le.ScanCallback
import android.bluetooth.le.ScanResult
import android.bluetooth.le.ScanSettings
import android.content.Context
import android.content.pm.PackageManager
import android.os.BatteryManager
import android.os.Build
import android.util.Log
import androidx.core.content.ContextCompat
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.softwiredtech.dashpilot.datamodel.dash.DASH_PREFS_NAME
import com.softwiredtech.dashpilot.datamodel.dash.DashState
import com.softwiredtech.dashpilot.datamodel.dash.DisplaySettings
import com.softwiredtech.dashpilot.datamodel.dash.PREF_ALWAYS_ON_BLIND_SPOT_MONITOR
import com.softwiredtech.dashpilot.datamodel.dash.PREF_DARK_MODE
import com.softwiredtech.dashpilot.datamodel.dash.PREF_DARK_MODE_BACKGROUND_GRAY
import com.softwiredtech.dashpilot.datamodel.dash.PREF_EXTRA_VEHICLE_BUS
import com.softwiredtech.dashpilot.datamodel.dash.PREF_RENDER_QUALITY
import com.softwiredtech.dashpilot.datamodel.dash.PREF_SHOW_CAR_BATTERY
import com.softwiredtech.dashpilot.datamodel.dash.PREF_SHOW_ODOMETER
import com.softwiredtech.dashpilot.datamodel.dash.PREF_SHOW_PHONE_BATTERY
import com.softwiredtech.dashpilot.datamodel.dash.PREF_USE_IMPERIAL
import com.softwiredtech.dashpilot.datamodel.dash.DEFAULT_ALWAYS_ON_BLIND_SPOT_MONITOR
import com.softwiredtech.dashpilot.datamodel.dash.DEFAULT_DARK_MODE
import com.softwiredtech.dashpilot.datamodel.dash.DEFAULT_DARK_MODE_BACKGROUND_GRAY
import com.softwiredtech.dashpilot.datamodel.dash.DEFAULT_EXTRA_VEHICLE_BUS
import com.softwiredtech.dashpilot.datamodel.dash.DEFAULT_RENDER_QUALITY
import com.softwiredtech.dashpilot.datamodel.dash.DEFAULT_SHOW_CAR_BATTERY
import com.softwiredtech.dashpilot.datamodel.dash.DEFAULT_SHOW_ODOMETER
import com.softwiredtech.dashpilot.datamodel.dash.DEFAULT_SHOW_PHONE_BATTERY
import com.softwiredtech.dashpilot.datamodel.dash.DEFAULT_USE_IMPERIAL
import com.softwiredtech.dashpilot.datamodel.dash.getPinnedControl
import com.softwiredtech.dashpilot.datamodel.dash.setPinnedControl
import com.softwiredtech.dashpilot.datamodel.dash.getFingerActions
import com.softwiredtech.dashpilot.datamodel.dash.setFingerActions
import com.softwiredtech.dashpilot.datamodel.dash.getWiperOffAutomation
import com.softwiredtech.dashpilot.datamodel.dash.setWiperOffAutomation
import com.softwiredtech.dashpilot.datamodel.dash.DEFAULT_CLIMATE_KEEP_MINUTES
import com.softwiredtech.dashpilot.datamodel.dash.getClimateKeepAutomation
import com.softwiredtech.dashpilot.datamodel.dash.setClimateKeepAutomation
import com.softwiredtech.dashpilot.datamodel.dash.getClimateKeepMinutes
import com.softwiredtech.dashpilot.datamodel.dash.setClimateKeepMinutes
import com.softwiredtech.dashpilot.ui.controls.controlById
import com.softwiredtech.dashpilot.datasource.DataSourceType
import com.softwiredtech.dashpilot.datasource.CommaDataSource
import com.softwiredtech.dashpilot.datasource.ConnectionStatus
import com.softwiredtech.dashpilot.datasource.IDataSource
import com.softwiredtech.dashpilot.datasource.DashKitBleManager
import com.softwiredtech.dashpilot.datasource.DashKitDataSource
import com.softwiredtech.dashpilot.datasource.WebsocketDataSource
import com.softwiredtech.dashpilot.ble.VehicleControl
import com.softwiredtech.dashpilot.jni.VehicleBridge
import com.softwiredtech.dashpilot.util.NetworkUtil
import com.softwiredtech.dashpilot.vehicle.CanFrameDecoder
import com.softwiredtech.dashpilot.vehicle.VehicleProfileLoader
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.combine
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.flow.flow
import kotlinx.coroutines.launch
import kotlinx.coroutines.suspendCancellableCoroutine
import kotlinx.coroutines.withTimeoutOrNull
import kotlin.coroutines.resume

class ConnectionViewModel(private var networkUtil: NetworkUtil) : ViewModel() {
    private val _dataSource = MutableStateFlow<IDataSource?>(null)
    private var connectionJob: Job? = null

    private val _connectionStatus = MutableStateFlow<ConnectionStatus>(ConnectionStatus.Disconnected)
    val connectionStatus = _connectionStatus.asStateFlow()

    private val _bleManager = MutableStateFlow<DashKitBleManager?>(null)
    val bleManager: StateFlow<DashKitBleManager?> = _bleManager.asStateFlow()

    // Data source the current/most recent session runs on. Drives which home
    // layout is shown (stat tiles, controls availability). Defaults to DashKit,
    // the primary source.
    private val _activeDataSource = MutableStateFlow(DataSourceType.DASHKIT)
    val activeDataSource = _activeDataSource.asStateFlow()

    var blePermissionGate: ((onGranted: () -> Unit) -> Unit)? = null

    private val _hasAutoNavigatedToDashboard = MutableStateFlow(false)
    val hasAutoNavigatedToDashboard = _hasAutoNavigatedToDashboard.asStateFlow()

    fun markAutoNavigated() {
        _hasAutoNavigatedToDashboard.value = true
    }

    private val _pinnedControl = MutableStateFlow<String?>(null)
    val pinnedControl = _pinnedControl.asStateFlow()

    fun loadPinnedControl(context: Context) {
        _pinnedControl.value = getPinnedControl(context)
    }

    fun togglePinnedControl(context: Context, id: String) {
        val next = if (_pinnedControl.value == id) null else id
        setPinnedControl(context, next)
        _pinnedControl.value = next
    }

    private val _wiperOffAutomation = MutableStateFlow(false)
    val wiperOffAutomation = _wiperOffAutomation.asStateFlow()

    private val _climateKeepAutomation = MutableStateFlow(false)
    val climateKeepAutomation = _climateKeepAutomation.asStateFlow()

    private val _climateKeepMinutes = MutableStateFlow(DEFAULT_CLIMATE_KEEP_MINUTES)
    val climateKeepMinutes = _climateKeepMinutes.asStateFlow()

    private val _fingerActions = MutableStateFlow<Map<Int, String>>(emptyMap())
    val fingerActions = _fingerActions.asStateFlow()

    fun loadAutomations(context: Context) {
        _wiperOffAutomation.value = getWiperOffAutomation(context)
        _climateKeepAutomation.value = getClimateKeepAutomation(context)
        _climateKeepMinutes.value = getClimateKeepMinutes(context)
        _fingerActions.value = getFingerActions(context)
    }

    fun updateWiperOffAutomation(context: Context, value: Boolean) {
        setWiperOffAutomation(context, value)
        _wiperOffAutomation.value = value
        pushWiperOff(value)
    }

    private fun pushWiperOff(enabled: Boolean) {
        _bleManager.value?.let { VehicleControl.sendWiperOff(it, enabled) }
    }

    fun updateClimateKeepAutomation(context: Context, value: Boolean) {
        setClimateKeepAutomation(context, value)
        _climateKeepAutomation.value = value
        _bleManager.value?.let { VehicleControl.sendClimateKeep(it, value) }
    }

    private var climateKeepMinutesPush: Job? = null

    fun updateClimateKeepMinutes(context: Context, value: Int) {
        setClimateKeepMinutes(context, value)
        _climateKeepMinutes.value = value
        // Debounced: the wheel picker fires once per detent while spinning.
        climateKeepMinutesPush?.cancel()
        climateKeepMinutesPush = viewModelScope.launch {
            delay(400)
            _bleManager.value?.let { VehicleControl.sendClimateKeepDuration(it, value) }
        }
    }

    fun setFingerAction(context: Context, fingers: Int, id: String?) {
        val next = _fingerActions.value.toMutableMap()
        if (id.isNullOrBlank()) next.remove(fingers) else next[fingers] = id
        _fingerActions.value = next
        setFingerActions(context, next)
        pushFingerAction(fingers, id)
    }

    fun changeFingerCount(context: Context, from: Int, to: Int) {
        if (from == to) return
        val current = _fingerActions.value
        val id = current[from] ?: return
        val next = current.toMutableMap()
        next.remove(from)
        next[to] = id
        _fingerActions.value = next
        setFingerActions(context, next)
        pushFingerAction(from, null)  // clear the old slot on the firmware
        pushFingerAction(to, id)
    }

    fun removeFingerAction(context: Context, fingers: Int) {
        setFingerAction(context, fingers, null)
    }

    private fun pushFingerAction(fingers: Int, id: String?) {
        val actionValue = controlById(id)?.gestureValue ?: VehicleControl.GESTURE_ACTION_NONE
        _bleManager.value?.let { VehicleControl.sendFingerAction(it, fingers, actionValue) }
    }

    enum class StartupRoute { LOADING, ONBOARDING_DASHKIT, AUTOCONNECT_DASHKIT, DEFAULT }

    // generation increments on every discovery run so the UI's routing effect
    // re-fires even when two consecutive runs resolve to the same route.
    data class StartupTarget(val route: StartupRoute, val generation: Int)

    private val _startupTarget = MutableStateFlow(StartupTarget(StartupRoute.LOADING, 0))
    val startupTarget = _startupTarget.asStateFlow()
    private var startupDiscoveryStarted = false
    private var discoveryGeneration = 0

    private var lastDataSourceType: String? = null
    private var lastServerAddress: String = ""

    // Data source the user explicitly chose this session (via the data source
    // menu). Restored on foreground instead of being overridden by discovery.
    private var userSelectedType: String? = null

    // Briefly scan for an advertising DashKit and decide the launch route:
    //  - found + bonded   -> auto-connect DashKit
    //  - found + unbonded -> open onboarding (pair flow)
    //  - not found / no permission -> default (setup + comma)
    fun runStartupDiscovery(context: Context, hasBlePermission: Boolean) {
        if (startupDiscoveryStarted) return
        startupDiscoveryStarted = true
        val generation = ++discoveryGeneration

        if (!hasBlePermission) {
            _startupTarget.value = StartupTarget(StartupRoute.DEFAULT, generation)
            return
        }

        viewModelScope.launch {
            val route = withTimeoutOrNull(STARTUP_SCAN_TIMEOUT_MS) {
                scanForDashKit(context)
            } ?: StartupRoute.DEFAULT
            _startupTarget.value = StartupTarget(route, generation)
        }
    }

    @SuppressLint("MissingPermission")
    private suspend fun scanForDashKit(context: Context): StartupRoute =
        suspendCancellableCoroutine { cont ->
            val adapter = (context.getSystemService(Context.BLUETOOTH_SERVICE) as? BluetoothManager)
                ?.adapter
            val scanner = adapter?.bluetoothLeScanner
            if (scanner == null) {
                if (cont.isActive) cont.resume(StartupRoute.DEFAULT)
                return@suspendCancellableCoroutine
            }

            val callback = object : ScanCallback() {
                override fun onScanResult(callbackType: Int, result: ScanResult) {
                    val name = result.device.name ?: result.scanRecord?.deviceName
                    if (name != DASHKIT_DEVICE_NAME) return
                    val bonded = result.device.bondState == BluetoothDevice.BOND_BONDED
                    try { scanner.stopScan(this) } catch (_: Exception) {}
                    if (cont.isActive) {
                        cont.resume(
                            if (bonded) StartupRoute.AUTOCONNECT_DASHKIT
                            else StartupRoute.ONBOARDING_DASHKIT
                        )
                    }
                }

                override fun onScanFailed(errorCode: Int) {
                    Log.e("ConnectionViewModel", "Startup scan failed: $errorCode")
                    if (cont.isActive) cont.resume(StartupRoute.DEFAULT)
                }
            }

            val settings = ScanSettings.Builder()
                .setScanMode(ScanSettings.SCAN_MODE_LOW_LATENCY)
                .build()
            try {
                scanner.startScan(null, settings, callback)
            } catch (e: Exception) {
                Log.e("ConnectionViewModel", "Startup scan could not start: ${e.message}")
                if (cont.isActive) cont.resume(StartupRoute.DEFAULT)
                return@suspendCancellableCoroutine
            }

            cont.invokeOnCancellation {
                try { scanner.stopScan(callback) } catch (_: Exception) {}
            }
        }

    private val _displaySettings = MutableStateFlow(DisplaySettings())
    private val _speedCameraDistance = MutableStateFlow(-1)

    private val _dashState = MutableStateFlow<Flow<DashState>?>(null)
    val dashState = _dashState.asStateFlow()

    private fun phoneBatteryFlow(context: Context): Flow<Int> = flow {
        val batteryManager = context.getSystemService(Context.BATTERY_SERVICE) as BatteryManager
        while (true) {
            val level = batteryManager.getIntProperty(BatteryManager.BATTERY_PROPERTY_CAPACITY)
            emit(level)
            delay(30_000)
        }
    }

    private fun hasBluetoothPermission(context: Context): Boolean {
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.S) return true
        return ContextCompat.checkSelfPermission(context, Manifest.permission.BLUETOOTH_CONNECT) ==
                PackageManager.PERMISSION_GRANTED &&
                ContextCompat.checkSelfPermission(context, Manifest.permission.BLUETOOTH_SCAN) ==
                PackageManager.PERMISSION_GRANTED
    }

    fun connect(
        context: Context,
        manualServerAddress: String,
        dataSourceType: String,
        userInitiated: Boolean = false,
    ) {
        if (dataSourceType == DataSourceType.DASHKIT && !hasBluetoothPermission(context)) {
            val gate = blePermissionGate
            if (gate != null) {
                gate { connect(context, manualServerAddress, dataSourceType, userInitiated) }
            } else {
                _connectionStatus.value = ConnectionStatus.Error("Missing BLE permission")
            }
            return
        }

        val current = _connectionStatus.value
        if (current !is ConnectionStatus.Disconnected && current !is ConnectionStatus.Error) return

        if (userInitiated) userSelectedType = dataSourceType
        lastDataSourceType = dataSourceType
        lastServerAddress = manualServerAddress
        _activeDataSource.value = dataSourceType

        var finalServerAddress = manualServerAddress
        _connectionStatus.value = ConnectionStatus.Connecting
        connectionJob = viewModelScope.launch(Dispatchers.IO) {
            val vehicleName = "tesla" // TODO: make configurable via UI
            val prefs = context.getSharedPreferences(DASH_PREFS_NAME, Context.MODE_PRIVATE)
            val extraBus = prefs.getBoolean(PREF_EXTRA_VEHICLE_BUS, DEFAULT_EXTRA_VEHICLE_BUS)
            val configFile = when (dataSourceType) {
                DataSourceType.DASHKIT -> "config_dashkit.json"
                else -> if (extraBus) "config_comma_extra_bus.json" else "config_comma_normal.json"
            }
            val profile = VehicleProfileLoader.loadProfile(context, vehicleName, configFile)
            val bridge = VehicleBridge()
            val ds = when (dataSourceType) {
                DataSourceType.DASHKIT -> {
                    val manager = DashKitBleManager(context)
                    _bleManager.value = manager
                    launch {
                        manager.connectionState.collect { st ->
                            if (st is ConnectionStatus.Error &&
                                _connectionStatus.value is ConnectionStatus.Connecting
                            ) {
                                _connectionStatus.value = st
                            }
                        }
                    }
                    val decoder = CanFrameDecoder(bridge, profile)
                    DashKitDataSource(manager, decoder)
                }
                DataSourceType.WEBSOCKET -> WebsocketDataSource()
                else -> CommaDataSource(bridge, profile)
            }

            if (finalServerAddress.isEmpty() && dataSourceType == DataSourceType.COMMA) {
                var localIp = networkUtil.getWifiIpAddress()
                while (localIp == null) {
                    delay(1000)
                    localIp = networkUtil.getWifiIpAddress()
                }
                finalServerAddress = networkUtil.findCanPublisher(initialIp = localIp)
            }

            _dataSource.value = ds

            _displaySettings.value = DisplaySettings(
                showPhoneBattery = prefs.getBoolean(PREF_SHOW_PHONE_BATTERY, DEFAULT_SHOW_PHONE_BATTERY),
                showCarBattery = prefs.getBoolean(PREF_SHOW_CAR_BATTERY, DEFAULT_SHOW_CAR_BATTERY),
                showOdometer = prefs.getBoolean(PREF_SHOW_ODOMETER, DEFAULT_SHOW_ODOMETER),
                useImperial = prefs.getBoolean(PREF_USE_IMPERIAL, DEFAULT_USE_IMPERIAL),
                darkMode = prefs.getBoolean(PREF_DARK_MODE, DEFAULT_DARK_MODE),
                alwaysOnBlindSpotMonitor = prefs.getBoolean(PREF_ALWAYS_ON_BLIND_SPOT_MONITOR, DEFAULT_ALWAYS_ON_BLIND_SPOT_MONITOR),
                renderQuality = prefs.getInt(PREF_RENDER_QUALITY, DEFAULT_RENDER_QUALITY),
                darkModeBackgroundGray = prefs.getInt(PREF_DARK_MODE_BACKGROUND_GRAY, DEFAULT_DARK_MODE_BACKGROUND_GRAY),
            )

            val combined = combine(
                ds.incomingMessages,
                phoneBatteryFlow(context),
                _displaySettings,
                _speedCameraDistance
            ) { carState, battery, settings, camDist ->
                val converted = if (settings.useImperial) carState.toImperial() else carState
                DashState(
                    carState = converted,
                    phoneBattery = battery,
                    currentTime = System.currentTimeMillis(),
                    displaySettings = settings,
                    speedCameraDistance = camDist
                )
            }
            _dashState.value = combined

            launch(Dispatchers.IO) { ds.connect(finalServerAddress) }
            ds.incomingMessages.first()
            _connectionStatus.value = ConnectionStatus.Connected

            // The firmware also persists the bindings in NVS, but re-sync every
            // finger count on each connect so its state matches the app exactly
            // (e.g. after a DashKit reboot or an edit made while disconnected).
            // The GATT services are discovered by the time the first CAN frame
            // arrives above.
            _bleManager.value?.let { mgr ->
                val actions = getFingerActions(context)
                _fingerActions.value = actions
                for (fingers in 3..5) {
                    val actionValue = controlById(actions[fingers])?.gestureValue
                        ?: VehicleControl.GESTURE_ACTION_NONE
                    VehicleControl.sendFingerAction(mgr, fingers, actionValue)
                }

                // Re-sync the wiper-off automation flag for the same reason.
                val wiperOff = getWiperOffAutomation(context)
                _wiperOffAutomation.value = wiperOff
                VehicleControl.sendWiperOff(mgr, wiperOff)

                // Re-sync the climate-keep automation the same way.
                val climateKeep = getClimateKeepAutomation(context)
                _climateKeepAutomation.value = climateKeep
                VehicleControl.sendClimateKeep(mgr, climateKeep)
                val climateKeepMin = getClimateKeepMinutes(context)
                _climateKeepMinutes.value = climateKeepMin
                VehicleControl.sendClimateKeepDuration(mgr, climateKeepMin)
            }
        }
    }

    fun updateDisplaySettings(settings: DisplaySettings) {
        _displaySettings.value = settings
    }

    fun bindSpeedCamera(flow: StateFlow<Pair<*, Int>?>) {
        viewModelScope.launch {
            flow.collect { _speedCameraDistance.value = it?.second ?: -1 }
        }
    }

    private fun teardownConnection() {
        connectionJob?.cancel()
        connectionJob = null
        _dataSource.value?.disconnect()
        _dataSource.value = null
        _bleManager.value?.disconnect()
        _bleManager.value = null
        _dashState.value = null
        _hasAutoNavigatedToDashboard.value = false
    }

    fun disconnect() {
        val wasConnecting = _connectionStatus.value is ConnectionStatus.Connecting
        teardownConnection()
        _connectionStatus.value = if (wasConnecting) {
            ConnectionStatus.Error("Discovery cancelled")
        } else {
            ConnectionStatus.Disconnected
        }
    }

    fun onAppForegrounded(context: Context, hasBlePermission: Boolean) {
        // Restart the keepalive before the firmware's timeout evicts a
        // surviving link.
        _bleManager.value?.onAppForegrounded()

        // First onStart after a cold launch (or while the permission dialog is
        // up): startup discovery is still deciding the route — let it finish.
        if (_startupTarget.value.route == StartupRoute.LOADING) return

        val type = lastDataSourceType
        if (type != null && (type == DataSourceType.DASHKIT || type == userSelectedType)) {
            if (type == DataSourceType.DASHKIT) {
                // The BLE link survives backgrounding. If it is still up (or
                // the manager is already connecting/auto-reconnecting), keep
                // it: a teardown + rescan here would drop a healthy link, and
                // rescanning on every foreground burns through Android's
                // 5-scans-per-30s budget, after which scans silently return
                // nothing and the app hangs in "Connecting…".
                val bleState = _bleManager.value?.connectionState?.value
                if (bleState == ConnectionStatus.Connected ||
                    bleState == ConnectionStatus.Connecting
                ) {
                    return
                }
            }
            // A dead DashKit session, or a source the user picked by hand:
            // restore it directly instead of re-running discovery.
            reconnect(context, type)
            return
        }

        if (!hasBlePermission) {
            // Can't scan for DashKit; just restore whatever ran before.
            if (type != null) reconnect(context, type)
            return
        }

        // No DashKit session and no explicit user choice — the app may have
        // been opened away from the car, so decide the route again exactly
        // like a fresh launch: auto-connect a bonded DashKit, onboard an
        // unbonded one, otherwise fall back to the default source.
        teardownConnection()
        _connectionStatus.value = ConnectionStatus.Disconnected
        startupDiscoveryStarted = false
        _startupTarget.value = StartupTarget(StartupRoute.LOADING, discoveryGeneration)
        runStartupDiscovery(context, hasBlePermission)
    }

    fun onAppBackgrounded() {
        _bleManager.value?.onAppBackgrounded()
    }

    private fun reconnect(context: Context, type: String) {
        val address = lastServerAddress
        teardownConnection()
        _connectionStatus.value = ConnectionStatus.Disconnected
        connect(context, address, type)
    }

    companion object {
        private const val DASHKIT_DEVICE_NAME = "DashKit"
        private const val STARTUP_SCAN_TIMEOUT_MS = 3000L
    }
}

File: dashpilot-ios/dashpilot/ViewModel/ConnectionViewModel.swift
import Foundation
import Observation
import UIKit

@Observable
final class ConnectionViewModel {

    /// UserDefaults key for the "onboarding was finished or skipped" flag
    /// (iOS stand-in for the Android bond-state check — iOS cannot ask
    /// whether the DashKit is already paired).
    static let onboardingCompletedKey = "dashkitOnboardingCompleted"

    /// Set when the launch scan finds a DashKit but onboarding has not been
    /// completed yet (or when the user replays onboarding from Settings).
    /// The app root observes it to present the onboarding cover; dismissing
    /// the cover clears it.
    var onboardingRequested = false

    private(set) var connectionStatus: ConnectionStatus = .disconnected
    private(set) var discoveredAddress: String?
    private(set) var discoveryError: String?

    /// Most recent state of the session, replayed to new subscribers so a view
    /// opened mid-session renders immediately.
    private(set) var latestDashState: DashState?

    /// Live per-view subscriptions. AsyncStream is single-consumer and dies
    /// with its consuming task, so every view gets its own stream and the pump
    /// fans out to all of them (the Android equivalent is a shared Flow).
    @ObservationIgnored
    private var dashSubscribers: [UUID: AsyncStream<DashState>.Continuation] = [:]

    /// Live BLE manager while a DashKit session is active. Views use it to
    /// send control commands and the settings screen for firmware/OTA access.
    private(set) var bleManager: DashKitBleManager?

    /// Source of the current (or last) session; stays set after a DashKit
    /// error so foregrounding retries the same source, like Android.
    private(set) var activeSourceType: DataSourceType?

    private var dataSource: (any IDataSource)?
    private var connectTask: Task<Void, Never>?
    private var startupDiscoveryStarted = false

    /// Returns a stream of `DashState` updates for one view. Each caller gets
    /// an independent stream, so one view leaving (its task being cancelled)
    /// tears down only its own subscription — not the pipeline feeding the
    /// other views. Subscriptions survive reconnects; the stream ends when the
    /// consuming task is cancelled.
    func dashStateStream() -> AsyncStream<DashState> {
        AsyncStream(bufferingPolicy: .bufferingNewest(1)) { continuation in
            let id = UUID()
            dashSubscribers[id] = continuation
            if let latest = latestDashState {
                continuation.yield(latest)
            }
            continuation.onTermination = { [weak self] _ in
                Task { @MainActor in
                    self?.dashSubscribers.removeValue(forKey: id)
                }
            }
        }
    }

    /// True when a new connection may start (mirrors Android: connect is
    /// allowed from Disconnected and Error).
    private var canStartConnection: Bool {
        switch connectionStatus {
        case .disconnected, .error: return true
        case .connecting, .connected: return false
        }
    }

    // MARK: - Startup discovery / warm launch

    /// Briefly scan for an advertising DashKit at launch (port of the Android
    /// startup discovery). If onboarding was already completed the DashKit is
    /// silently auto-connected; otherwise the onboarding flow is requested and
    /// its "Pair device" CTA drives the first connection.
    func startupAutoConnect() {
        guard !startupDiscoveryStarted else { return }
        startupDiscoveryStarted = true
        guard connectionStatus == .disconnected else { return }

        Task { @MainActor [weak self] in
            let found = await DashKitStartupScanner().scan(timeout: 3)
            guard let self, found, self.canStartConnection, self.dataSource == nil else { return }
            if UserDefaults.standard.bool(forKey: Self.onboardingCompletedKey) {
                self.connectDashKit()
            } else {
                self.onboardingRequested = true
            }
        }
    }

    /// Called when the app returns to the foreground. Keeps a healthy DashKit
    /// link (a teardown + rescan would drop it), revives a dead DashKit
    /// session, and otherwise re-runs startup discovery in case the app was
    /// opened away from the car before.
    func onAppForegrounded() {
        switch activeSourceType {
        case .dashkit:
            if let manager = bleManager,
               manager.connectionState == .connected || manager.connectionState == .connecting {
                return
            }
            disconnect()
            connectDashKit()
        case .comma, .websocket:
            // WiFi sources are user-initiated on iOS; leave them alone.
            return
        case nil:
            guard connectionStatus == .disconnected else { return }
            startupDiscoveryStarted = false
            startupAutoConnect()
        }
    }

    // MARK: - DashKit

    func connectDashKit() {
        guard canStartConnection else { return }
        if case .error = connectionStatus { disconnect() }  // drop failed-session remnants
        connectionStatus = .connecting
        discoveryError = nil
        discoveredAddress = nil
        activeSourceType = .dashkit

        let manager = DashKitBleManager()
        bleManager = manager
        manager.onStateChange = { [weak self] state in
            guard let self else { return }
            // Surface BLE failures while we are still waiting for the first
            // CAN frame so the UI leaves "Connecting…" (mirrors Android).
            if case .error = state, self.connectionStatus == .connecting {
                self.connectionStatus = state
            }
        }

        let ds = DashKitDataSource(manager: manager)
        dataSource = ds
        ds.connect(address: "")
        startStreaming(ds, resyncDashKit: true)
    }

    /// Pushes the persisted multi-finger bindings and the wiper-off flag to
    /// the firmware. The firmware persists them in NVS too, but re-syncing on
    /// each connect keeps its state matching the app exactly (e.g. after a
    /// DashKit reboot or an edit made while disconnected).
    func resyncDashKitAutomations() {
        guard let manager = bleManager else { return }
        let raw = UserDefaults.standard.string(forKey: "finger_actions") ?? ""
        let actions = AutomationsView.parseFingerActions(raw)
        let byCount = Dictionary(uniqueKeysWithValues: actions.map { ($0.fingerCount, $0.controlId) })
        for fingers in 3...5 {
            let actionValue = byCount[fingers].flatMap { controlById($0)?.gestureValue }
                ?? VehicleControl.gestureActionNone
            VehicleControl.sendFingerAction(manager, fingers: fingers, actionValue: actionValue)
        }
        let wiperOff = UserDefaults.standard.bool(forKey: "wiper_off_automation")
        VehicleControl.sendWiperOff(manager, enabled: wiperOff)
        let climateKeep = UserDefaults.standard.bool(forKey: "climate_keep_automation")
        VehicleControl.sendClimateKeep(manager, enabled: climateKeep)
        let climateKeepMinutes = UserDefaults.standard.integer(forKey: "climate_keep_minutes")
        VehicleControl.sendClimateKeepDuration(manager, minutes: climateKeepMinutes > 0 ? climateKeepMinutes : 5)
    }

    // MARK: - Comma (WiFi)

    /// Auto-discover a publisher on the local subnet, then connect.
    func autoConnect() {
        guard canStartConnection else { return }
        if case .error = connectionStatus { disconnect() }  // drop failed-session remnants
        connectionStatus = .connecting
        discoveryError = nil
        discoveredAddress = nil
        activeSourceType = .comma

        connectTask = Task { @MainActor [weak self] in
            guard let self else { return }

            // Retry discovery in a loop until found or cancelled
            var foundIp: String?
            while foundIp == nil && !Task.isCancelled {
                foundIp = await NetworkUtil.findPublisher(endpoint: "can", timeoutMs: 5000)
                if foundIp == nil && !Task.isCancelled {
                    // Brief pause before retrying
                    try? await Task.sleep(for: .milliseconds(700))
                }
            }

            guard !Task.isCancelled, let ip = foundIp else {
                if !Task.isCancelled {
                    self.discoveryError = "Could not find device on the network"
                    self.connectionStatus = .disconnected
                }
                return
            }

            self.discoveredAddress = ip
            self.startCommaDataSource(address: ip)
        }
    }

    /// Connect directly to a known address (manual entry).
    func connect(serverAddress: String) {
        guard canStartConnection else { return }
        if case .error = connectionStatus { disconnect() }  // drop failed-session remnants
        connectionStatus = .connecting
        discoveryError = nil
        discoveredAddress = serverAddress
        activeSourceType = .comma
        startCommaDataSource(address: serverAddress)
    }

    func disconnect() {
        connectTask?.cancel()
        connectTask = nil
        dataSource?.disconnect()
        dataSource = nil
        bleManager?.onStateChange = nil
        bleManager?.disconnect()
        bleManager = nil
        activeSourceType = nil
        latestDashState = nil
        discoveredAddress = nil
        discoveryError = nil
        connectionStatus = .disconnected
    }

    // MARK: - Private

    private func startCommaDataSource(address: String) {
        let extraBus = UserDefaults.standard.bool(forKey: DisplaySettings.keyExtraVehicleBus)
        let ds: CommaDataSource
        if extraBus {
            ds = CommaDataSource(
                vehicleType: "tesla_extra",
                dbcFiles: [
                    (1, "bus_1_tesla_vehicle"),
                    (2, "bus_2_tesla_party")
                ]
            )
        } else {
            ds = CommaDataSource()
        }
        dataSource = ds
        ds.connect(address: address)
        startStreaming(ds, resyncDashKit: false)
    }

    /// Consumes the data source's `CarState` stream, republishes it as
    /// `DashState` to every subscribed view, and flips to `.connected` on the
    /// first message. Subscribers are not finished when a session ends — a
    /// reconnect's pump simply resumes feeding them.
    private func startStreaming(_ ds: any IDataSource, resyncDashKit: Bool) {
        connectTask = Task { @MainActor [weak self] in
            guard let self else { return }
            UIDevice.current.isBatteryMonitoringEnabled = true
            var first = true
            for await carState in ds.incomingMessages {
                if first {
                    self.connectionStatus = .connected
                    first = false
                    // The GATT services are discovered by the time the first
                    // CAN frame arrives.
                    if resyncDashKit {
                        self.resyncDashKitAutomations()
                    }
                }
                let settings = DisplaySettingsState.fromUserDefaults()
                let converted = settings.useImperial ? carState.toImperial() : carState
                let batteryLevel = UIDevice.current.batteryLevel
                let state = DashState(
                    carState: converted,
                    displaySettings: settings,
                    phoneBattery: batteryLevel >= 0 ? Int(batteryLevel * 100) : -1,
                    currentTime: Int64(Date().timeIntervalSince1970 * 1000),
                    dataSourceType: self.activeSourceType
                )
                self.latestDashState = state
                for continuation in self.dashSubscribers.values {
                    continuation.yield(state)
                }
            }
        }
    }
}
